src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,217 @@
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            
+            var namespaceDeclarationCount = root.DescendantNodes().OfType<NamespaceDeclarationSyntax>().Count();
+
+            if (namespaceDeclarationCount == 1)
+            {
+                var token = root.FindToken(position);
+                if (token.Kind() == SyntaxKind.EndOfFileToken)
+                {
+                    token = token.GetPreviousToken();
+                }
+                // Should trigger if cursor is on the name of only namespace declaration in this document.
+                var namespaceDeclaration = token.GetAncestor<NamespaceDeclarationSyntax>();
+
+                var shouldTrigger = namespaceDeclaration != null 
+                    && namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.GetDiagnostics().All(diag => diag.Severity != DiagnosticSeverity.Error)
+                    && !ContainsPartialDeclaration(document, namespaceDeclaration, cancellationToken);
+                return (shouldTrigger, shouldTrigger ? namespaceDeclaration : null);
+            }
+
+            if (namespaceDeclarationCount == 0 && root is CompilationUnitSyntax compilationUnit)
+            {
+                // Should trigger if cursor is on the name of first member declaration in the 
+                // compilation unit when there's no namespace decalration in the document.
+                var firstMemberDeclaration = compilationUnit.Members.FirstOrDefault();
+                return (firstMemberDeclaration != null 
+                    && firstMemberDeclaration.GetNameToken().Span.IntersectsWith(position) 
+                    && !ContainsPartialDeclaration(document, compilationUnit, cancellationToken), null);
+            }
+
+            return (false, null);
+        }
+
+        private bool ContainsPartialDeclaration(Document document, SyntaxNode node, CancellationToken cancellationToken = default)
+        {
+            // This is just a quick check for `partial` keyword.
+            var memberDeclarations = GetMemberDeclarationsInContainer(node);
+            foreach (TypeDeclarationSyntax declaration in memberDeclarations.Where(decl => decl is TypeDeclarationSyntax))
+            {
+                if (declaration.Modifiers.Any(token => token.IsKind(SyntaxKind.PartialKeyword)))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        protected override SyntaxNode ChangeNamespaceDeclaration(SyntaxNode root, ImmutableArray<string> declaredNamespaceParts, ImmutableArray<string> targetNamespaceParts)
+        {
+            if (declaredNamespaceParts.IsDefault || targetNamespaceParts.IsDefault || !(root is CompilationUnitSyntax compilationUnit))
```

Comment:
when is the root ever *not* a compilation unit?

Reply:
This is also used to do the conversion.

Reply:
Fixed